target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_seil.c

```diff
@@ -0,0 +1,204 @@
+#include <linux/vmalloc.h>
+#include <linux/of.h>
+#include <linux/mtd/mtd.h>
+#include <linux/mtd/partitions.h>
+#include <linux/byteorder/generic.h>
+
+#include "mtdsplit.h"
+
+#define NR_PARTS		2
+#define SEIL_VFMT		1
+#define LDR_ENV_PART_NAME	"bootloader-env"
+#define LDR_ENV_KEY_BOOTDEV	"BOOTDEV"
+
+struct seil_header {
+	uint64_t	id;		/* Identifier */
+	uint8_t		copy[80];	/* Copyright */
+	uint32_t	dcrc;		/* Data CRC Checksum */
+	uint32_t	vfmt;		/* Image Version Format */
+	uint32_t	vmjr;		/* Image Version Major */
+	uint32_t	vmnr;		/* Image Version Minor */
+	uint8_t		vrel[32];	/* Image Version Release */
+	uint32_t	dxor;		/* xor value for Data? */
+	uint32_t	dlen;		/* Data Length */
+};
+
+/*
+ * check whether the current mtd device is active or not
+ *
+ * example of BOOTDEV value (IIJ SA-W2):
+ *   - "flash"   : primary image on flash
+ *   - "rescue"  : secondary image on flash
+ *   - "usb"     : usb storage
+ *   - "lan0/1"  : network
+ */
+static bool seil_bootdev_is_active(struct device_node *np)
+{
+	struct mtd_info *env_mtd = get_mtd_device_nm(LDR_ENV_PART_NAME);
+	char *buf, *var, *value, *eq;
+	const char *devnm;
+	size_t rdlen;
+	int ret;
+
+	/*
+	 * read bootdev name of the partition
+	 * if doesn't exist, return true and skip checking of active device
+	 */
+	ret = of_property_read_string(np, "iij,bootdev-name", &devnm);
+	if (ret == -EINVAL)
+		return true;
+	else if (ret < 0)
+		return false;
+
+	if (IS_ERR(env_mtd)) {
+		pr_err("failed to get mtd device \"%s\"", LDR_ENV_PART_NAME);
+		return false;
+	}
+
+	buf = vmalloc(env_mtd->size);
+	if (!buf)
+		return false;
+
+	ret = mtd_read(env_mtd, 0, env_mtd->size, &rdlen, buf);
+	if (ret || rdlen != env_mtd->size) {
+		pr_err("failed to read from mtd (%d)\n", ret);
+		goto exit_vfree;
+	}
+
+	for (var = buf, ret = false;
+	     var < buf + env_mtd->size && *var;
+	     var = value + strlen(value) + 1) {
+		eq = strchr(var, '=');
+		if (!eq)
+			break;
+		*eq = '\0';
+		value = eq + 1;
+
+		pr_debug("ENV: %s=%s\n", var, value);
+		if (!strcmp(var, LDR_ENV_KEY_BOOTDEV)) {
+			ret = !strcmp(devnm, value);
+			break;
+		}
+	}
+
+exit_vfree:
+	vfree(buf);
+
+	return ret;
+}
+
+static int mtdsplit_parse_seil_fw(struct mtd_info *master,
+				  const struct mtd_partition **pparts,
+				  struct mtd_part_parser_data *data)
+{
+	struct device_node *np = mtd_get_of_node(master);
+	struct mtd_partition *parts;
+	struct seil_header *header;
+	u_char *buf;
+	size_t image_size = 0;
+	size_t rootfs_offset;
+	size_t buflen;
+	size_t retlen;
+	enum mtdsplit_part_type type;
+	int ret;
+	u64 id;
+
+	if (!seil_bootdev_is_active(np))
+		return -ENODEV;
+
+	ret = of_property_read_u64(np, "iij,seil-id", &id);
+	if (ret) {
+		pr_err("failed to get iij,seil-id from dt\n");
+		return ret;
+	}
+	pr_debug("got seil-id=0x%016llx from dt\n", id);
+
+	parts = kcalloc(NR_PARTS, sizeof(*parts), GFP_KERNEL);
+	if (!parts)
+		return -ENOMEM;
+
+	buflen = sizeof(struct seil_header);
+	buf = vmalloc(buflen);
```

Comment:
`buf` is about 140 bytes big, I would prefer `kmalloc()` here.

Reply:
I used the same since mtdsplit_uimage is using ```vmalloc``` for header.
But indeed, the sizes of those headers are small enough, so ```kmalloc``` is better.

Reply:
I mage an additional change to read header directly without buf:
https://github.com/musashino-build/openwrt/commit/f1a387376dde862019ef21416e30db29cdb2d65a

Should this change also be applied?

Reply:
This looks fine. I think the `(void *)` is not needed.

Reply:
Thank you. I tried to removing ```(void *)```, but failed to compile due to ```incompatible pointer type```.

```
drivers/mtd/mtdsplit/mtdsplit_seil.c: In function 'mtdsplit_parse_seil_fw':
drivers/mtd/mtdsplit/mtdsplit_seil.c:128:52: error: passing argument 5 of 'mtd_read' from incompatible pointer type [-Werror=incompatible-pointer-types]
  128 |         ret = mtd_read(master, 0, hdrlen, &retlen, &header);
      |                                                    ^~~~~~~
      |                                                    |
      |                                                    struct seil_header *
In file included from drivers/mtd/mtdsplit/mtdsplit_seil.c:11:
./include/linux/mtd/mtd.h:508:22: note: expected 'u_char *' {aka 'unsigned char *'} but argument is of type 'struct seil_header *'
  508 |              u_char *buf);
      |              ~~~~~~~~^~~
```

I'll apply https://github.com/musashino-build/openwrt/commit/f1a387376dde862019ef21416e30db29cdb2d65a without removing ```(void *)``` and update this PR...

Reply:
ok makes sense. I thought it gets a `void *`, but it gets a `u_char *`. 
The cast is fine.